File: 269.alien-dictionary.cpp
/*
 * @lc app=leetcode id=269 lang=cpp
 *
 * [269] Alien Dictionary
 */
#include <algorithm>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
public:
    // T: O(V+E) = O(m*n + m*n) = O(m*n), m := len(words), n := max(len(w) for w in words)
    // S: O(m*n)
    string alienOrder(vector<string>& words) {
        // Build graph
        unordered_map<char, unordered_set<char>> graph;
        unordered_map<char, int> inDeg;
        for (const auto& w : words) {
            for (const auto& c : w) {
                // IMPORTANT: initialize for all chars 0 in-deg to distinguish from non-existent chars
                inDeg[c] = 0; 
            }
        }
        for (int i = 1; i < words.size(); ++i) {
            string pre = words[i - 1], cur = words[i];
            int n = min(pre.size(), cur.size());
            
            // By comparing the adjacent words, the first character that is different between 
            // two adjacent words reflect the lexicographical order
            for (int j = 0; j < n; ++j) {
                if (pre[j] != cur[j]) {
                    graph[pre[j]].insert(cur[j]);

                    // IMPORTANT: we may update inDeg here rather than below, but need to check 
                    // if graph[pre[j]] already contains cur[j]
                    
                    break;
                }

                // The word order is probably incorrect, because when pre[:n] == cur[:n] but len(pre) > len(cur),
                // namely cur is a prefix of pre, then lexicographically pre > cur. Impossible to give a solution.
                if (j == n - 1 && pre.size() > cur.size())  return "";
            }
        }
        for (const auto& [_, neighbors] : graph) {
            for (const auto& u : neighbors) {
                ++inDeg[u];
            }   
        }

        // Topological sort
        queue<char> q;
        for (const auto& [u, deg] : inDeg) {
            if (deg == 0)  q.push(u);
        }
        string res;
        res.reserve(inDeg.size());
        while (q.size()) {
            char u = q.front(); q.pop();
            res.push_back(u);

            auto it = graph.find(u);
            if (it != graph.end()) {
                for (const auto &v : it->second) {
                    if (--inDeg[v] == 0) {
                        q.push(v);
                    }
                }
            }
        }
        // If any cycles exist, then it's impossible to form a solution
        return res.size() == inDeg.size() ? res : "";
    }
};
// @lc code=end

File: 269.alien-dictionary.py
#
# @lc app=leetcode id=269 lang=python3
#
# [269] Alien Dictionary
#
from collections import deque
from typing import *

# @lc code=start
A = ord('a')

def cid(c):
    return ord(c) - A

class Solution:
    # Graph BFS + topological sort
    # Build a DAG graph so that we can get a possible order without invalid dependencies.
    # T: O(n * l + s ^ 2), n := len(words), l := max word length, s := 26
    # S: O(s ^ 2)
    def alienOrder(self, words: List[str]) -> str:
        graph = [[False] * 26 for _ in range(26)]
        in_deg = [-1 for _ in range(26)]  # default to -1 for non-existent chars
        n_uniq_chars = 0

        for w in words:
            for c in w:
                c = cid(c)
                if in_deg[c] == -1:
                    in_deg[c] = 0
                    n_uniq_chars += 1

        # Build a graph by checking each adjacent pairs of words 
        # (should be sorted lexicographically in the Alien language) 
        for i in range(1, len(words)):
            w1 = words[i - 1]
            w2 = words[i]
            n1 = len(w1)
            n2 = len(w2)
            n = min(n1, n2)
            for j in range(n):
                # compare chars at same index
                if w1[j] != w2[j]:
                    c1 = cid(w1[j])
                    c2 = cid(w2[j])
                    
                    # Form an edge from c1 to c2
                    if not graph[c1][c2]:
                        graph[c1][c2] = True
                        in_deg[c2] += 1
                    
                    break
                
                if j == n - 1 and n1 > n2:
                    # Invalid order
                    return ""
        
        # Topological sort to detect cycle
        res = []
        q = deque()
        for i in range(len(in_deg)):
            if in_deg[i] == 0:
                q.append(i)

        while len(q) > 0:
            i = q.popleft()
            res.append(chr(i + A))
            for j in range(len(graph[i])):
                if graph[i][j]:
                    in_deg[j] -= 1
                    if in_deg[j] == 0:
                        q.append(j)

        # Cycle exists if len(res) != n_uniq_chars
        # The result is the lexicographically increasing order of the chars in this language.
        return "".join(res) if len(res) == n_uniq_chars else ""
    
    # Similar problems: given some candidate subsequences, check if they are drawn 
    # from the same sequence with unique numbers

# @lc code=end

